warehouse/cache/origin/fastly.py

```diff
@@ -45,6 +45,7 @@ def __init__(self, *, api_key, service_id, purger):
         self.api_key = api_key
         self.service_id = service_id
         self._purger = purger
+        self.keys = set()
```

Comment:
I think this violates the API of the ``IOriginCache`` a bit, since technically a single ``IOriginCache`` can live through multiple requests (although the current implementation does not do that). This is alluded to by the fact that the ``cache()`` method requires passing in the ``request`` object.

I think this is along the right path, but I think probably you should just hang this list off of the request object somewhere. Another option is to store a dictionary of ``request: list(keys)``. This would likely need to be a ``weakref.WeakKeyDictionary`` to prevent having a memory leak.